### PR TITLE
search: Add operators to search reference

### PR DESCRIPTION
**Pull request by fkling**
I refactored the types so that `SearchReferenceEntry` can work with both kinds of entries. I decided to use two separate types since the information is slightly different. I'm not sure it's worth the effort to consolidate them into a single one, but I'm open to suggestions.

### CODE REVIEW

File: client/web/src/search/results/sidebar/SearchReference.tsx [hunk 1]
```diff
@@ -235,20 +245,47 @@ To use this filter, the search query must contain \`type:diff\` or \`type:commit
...
+const commonFilters = filterInfos
     .filter(info => info.commonRank !== undefined)
     // commonRank will never be undefined here, but TS doesn't seem to know
     .sort((a, b) => (a.commonRank as number) - (b.commonRank as number))
 
+const operatorInfo: OperatorInfo[] = [
+    {
+        operator: 'AND',
+        alias: 'and',
+        placeholder: parsePlaceholder('{a} AND {b}'),
```
Comment (fkling) [inline client/web/src/search/results/sidebar/SearchReference.tsx:261]: These probably need better placeholders, maybe `term AND term` or `expression AND expression`.
I feel like the current list (see screenshot) looks a bit sad.

File: client/web/src/search/results/sidebar/SearchReference.tsx [hunk 2]
```diff
@@ -268,7 +268,7 @@ const operatorInfo: OperatorInfo[] = [
         alias: 'or',
         placeholder: parsePlaceholder('{a} OR {b}'),
         description:
-            'Returns file content matching either on the left or right side, or both (set union). The number of results reports the number of matches of both strings. Note the regex or operator `|` may not work as expected with certain operators for example `file:(internal/repos)|(internal/gitserver)`, to recieve the expected results use [subexpressions](https://docs.sourcegraph.com/code_search/tutorials/search_subexpressions), `(file:internal/repos or file:internal/gitserver)`',
```
Comment (rvantonder) [inline client/web/src/search/results/sidebar/SearchReference.tsx]: Works for me to remove this. Still waiting on clarification on what this intended to communicate, since there was clearly some snag for someone using this.

File: client/web/src/search/results/sidebar/SearchReference.tsx (hunk 1 above)
Comment (rvantonder) [reply client/web/src/search/results/sidebar/SearchReference.tsx:261]: Maybe `expr` also. Whatever feels good to you here. I'm OK with `a and b` or something a little bit more indicative.

File: client/web/src/search/results/sidebar/SearchReference.tsx [hunk 3]
```diff
@@ -235,20 +245,47 @@ To use this filter, the search query must contain \`type:diff\` or \`type:commit
...
+        description:
+            'Returns file content matching either on the left or right side, or both (set union). The number of results reports the number of matches of both strings.',
+        examples: ['conf.Get( or log15.Error(', 'conf.Get( OR log15.Error( OR after'],
+    },
+    {
+        operator: 'NOT',
+        alias: 'not',
+        placeholder: parsePlaceholder('NOT {a}'),
+        description:
+            '`NOT` can be used in place of `-` to negate keywords, such as `file`, `content`, `lang`, `repohasfile`, and `repo`. For search patterns, `NOT` excludes documents that contain the term after `NOT`. For readability, you can also include the `AND` operator before a `NOT` (i.e. `panic NOT ever` is equivalent to `panic AND NOT ever`).',
```
Comment (rvantonder) [inline client/web/src/search/results/sidebar/SearchReference.tsx:279]: Just seems to read a bit long maybe in this context of the reference panel. I think the reason why the docs talk about using `not` in place of `-` is because that syntax, like `-file` appear before the operator section. So basically, I think we can shorten this to something like "`NOT` can be prepended to negate filters like `file`, `lang`, `repo`. Prepending `NOT` to search patterns excludes documents that contain the pattern. For readability, you may use `NOT` in conjunction with `AND` if you like: <example>".

If something like that sounds better to you just go ahead with the change.

File: client/web/src/search/results/sidebar/SearchReference.tsx [hunk 4]
```diff
@@ -235,20 +245,47 @@ To use this filter, the search query must contain \`type:diff\` or \`type:commit
...
+        alias: 'and',
+        placeholder: parsePlaceholder('{a} AND {b}'),
+        description:
+            'Returns results for files containing matches on the left and right side of the `and` (set intersection).',
+        examples: ['conf.Get( and log15.Error(', 'conf.Get( AND log15.Error( AND after'],
+    },
+    {
+        operator: 'OR',
+        alias: 'or',
+        placeholder: parsePlaceholder('{a} OR {b}'),
```
Comment (rvantonder) [inline client/web/src/search/results/sidebar/SearchReference.tsx:269]: In my experience, people are prone to trip themselves up over syntax (what happens when I type `repo:foo a or b` what does the `or` apply to? For this reason I think it would be sensible to add parentheses to the placeholder by default. Your thoughts?

```suggestion
        placeholder: parsePlaceholder('({a} OR {b})'),
```

File: client/web/src/search/results/sidebar/SearchReference.tsx [hunk 5]
```diff
@@ -424,6 +465,15 @@ const SearchReferenceExample: React.FunctionComponent<SearchReferenceExampleProp
...
                             )
+                        case 'keyword':
+                            // We are using example.slice instead of term.value
+                            // to get the actual character sequence in the
+                            // example. term.value doesn't preserve case
+                            return (
+                                <span className="search-filter-keyword">
+                                    {example.slice(term.range.start, term.range.end)}
+                                </span>
+                            )
```
Comment (rvantonder) [inline client/web/src/search/results/sidebar/SearchReference.tsx:476]: Hmmm, I'm surprised to see this says `.value` doesn't preserve keyword case--the scanner code does appear to do so:

https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/shared/src/search/query/scanner.ts?L315&subtree=true

I also tested this with a quick snapshot test, and indeed the `and` operators are cased appropriately

```json
{
   "type":"success",
   "term":[
// ... stuff
       {
         "type":"keyword",
         "value":"AND",
         "range":{
            "start":2,
            "end":5
         },
         "kind":"and"
      },
// ...stuff
      {
         "type":"keyword",
         "value":"and",
         "range":{
            "start":8,
            "end":11
         },
         "kind":"and"
      },
   ]
}
```

Perhaps you were looking at `token.kind`? (which indeed does not preserve case). Could you double check `token.value` really doesn't work?

**Review by rvantonder — APPROVED** [review]
Nice. left some non-blocking comments

**Comment by rvantonder** [comment]
Oh yeah, you'll need to approve the percy UI check (it's a legitimate change, you just approve this addition). You should have access I think.

File: client/web/src/search/results/sidebar/SearchReference.tsx (hunk 3 above)
Comment (fkling) [reply client/web/src/search/results/sidebar/SearchReference.tsx:279]: Sounds good to me. Do you think people understand the difference between "filter" and "pattern"?

File: client/web/src/search/results/sidebar/SearchReference.tsx (hunk 4 above)
Comment (fkling) [reply client/web/src/search/results/sidebar/SearchReference.tsx:269]: Yep, I like that.

File: client/web/src/search/results/sidebar/SearchReference.tsx (hunk 5 above)
Comment (fkling) [reply client/web/src/search/results/sidebar/SearchReference.tsx:476]: You are right, I think I was looking at the wrong example query.

File: client/web/src/search/results/sidebar/SearchReference.tsx (hunk 3 above)
Comment (rvantonder) [reply client/web/src/search/results/sidebar/SearchReference.tsx]: I think there's a genuine need to describe usage of these in terms of separate things (a filter is qualified by something like `file`, and search pattern is just an unqualified string). The terms "filter" versus "pattern" seem the most straightforward to me, from a linguistic perspective, but if you have other ideas go for it. The question on how to surface this better, I'm not sure besides inlining more examples in the phrasing like "... filters like `file:`, `lang:`, `repo:`. Prepending `NOT` to search terms (like `foo`)...."

**Review by limitedmage — APPROVED** [review]
🚀

File: client/web/src/search/results/sidebar/SearchReference.tsx (hunk 3 above)
Comment (fkling) [reply client/web/src/search/results/sidebar/SearchReference.tsx]: To be clear, I'm not implying that he difference shouldn't be explained/exist. Using correct terminology is important. I guess I'm wondering whether we do enough teach about the terminology... but that's an entirely separate issue.

File: CHANGELOG.md [hunk 6]
```diff
@@ -16,6 +16,7 @@ All notable changes to Sourcegraph are documented in this file.
 ### Added
 
 - Code Insights background queries now process in a priority order backwards through time. This will allow insights to populate concurrently. [#23101](https://github.com/sourcegraph/sourcegraph/pull/23101)
+- Operator documentation has been added to the Search Reference sidebar section. [#23116]https://github.com/sourcegraph/sourcegraph/pull/23116)
```
Comment (lguychard) [inline CHANGELOG.md:19]: @fkling FYI this link is missing an opening `(` 😉

Comment (fkling) [reply CHANGELOG.md:19]: :(

File: client/web/src/search/results/sidebar/SearchReference.tsx [hunk 7]
```diff
@@ -424,6 +466,8 @@ const SearchReferenceExample: React.FunctionComponent<SearchReferenceExampleProp
                                     {term.value?.quoted ? `"${term.value.value}"` : term.value?.value}
                                 </React.Fragment>
                             )
+                        case 'keyword':
+                            return <span className="search-filter-keyword">{term.value}</span>
```
Comment (vovakulikov) [inline client/web/src/search/results/sidebar/SearchReference.tsx:470]: @fkling I think we forgot to add key prop here

Comment (fkling) [reply client/web/src/search/results/sidebar/SearchReference.tsx:470]: Thank you @vovakulikov for pointing that out!

Comment (vovakulikov) [reply client/web/src/search/results/sidebar/SearchReference.tsx:470]: I'm glad to help 😉
